### PR TITLE
feature/mariadb-upgrade-10.2

### LINKED ISSUE #5907: Upgrade MariaDB
Support end will in October.

https://mariadb.com/kb/en/changes-improvements-in-mariadb-101/
> The first stable release was in October 2014, and it will be supported until October 2020.

### PR DESCRIPTION
**Pull request by jrouzierinverse**
# Description
Upgrade to MariaDB

# NEW Package(s) required
MariaDB-server-10.2.37
MariaDB-client-10.2.37

# Issue
fixes #5907

# Delete branch after merge
YES

# Checklist
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## Enhancements
* Upgrade to MariaDB 10.2.37

# UPGRADE file entries
## MariaDB upgrade (CentOS + RHEL only)

Before packetfence upgrade:

In order to upgrade the MariaDB metadata files and tables, first stop any started process.

  systemctl stop packetfence-mariadb
  rpm -e --nodeps MariaDB-client MariaDB-common MariaDB-server MariaDB-shared
  yum install --enablerepo=packetfence MariaDB-server-10.2.37

Then start a mysqld_safe process manually (this will start a background process)
 
   systemctl unmask mariadb
   systemctl start mariadb

Then, execute the upgrade script and enter the root password when prompted

  mysql_upgrade -u root -p

When done, stop mariadb

 systemctl stop mariadb
 systemctl mask mariadb
 
Note that it might take up to a few minutes for the process to exit depending on the size of your database.

Once done, restart the MariaDB service (managed by PacketFence)

  systemctl start packetfence-mariadb

Next proceed with yum update packetfence

## Centos7 Cluster upgrade

```
tee /etc/yum.repos.d/mariadb.repo << EOF
[mariadb]
name = MariaDB
enabled=0
baseurl = http://yum.mariadb.org/10.2/centos7-amd64
gpgkey=https://yum.mariadb.org/RPM-GPG-KEY-MariaDB
gpgcheck=1
EOF
```

bin/pfcmd service pf stop
systemctl stop packetfence-mariadb
rpm -e --nodeps MariaDB-client MariaDB-common MariaDB-server MariaDB-shared

yum install --enablerepo=mariadb MariaDB-server MariaDB-backup

systemctl unmask mariadb
systemctl start mariadb

if error with "Recovering after a crash using tc.log" then delete tc.log

mysql_upgrade -u root -p

systemctl stop mariadb
systemctl mask mariadb

yum update packetfence --enablerepo=packetfence

bin/pfcmd generatemariadbconfig

/usr/local/pf/sbin/pf-mariadb --force-new-cluster

## Debian 9 only

Before packetfence upgrade:

dpkg -r --force-depends mariadb-server mariadb-client-10.1 mariadb-client-core-10.1 mariadb-common mariadb-server-10.1 mariadb-server-core-10.1 libmariadbclient18

apt  install mariadb-server-10.2 mariadb-common mariadb-client-10.2 mariadb-client-core-10.2 mariadb-server-core-10.2 libmariadb3 libmariadbclient18 mariadb-server

### CODE REVIEW

**Comment by extrafu** [comment]
@fdurand and @lzammit and @JeGoi - please test that PR on a standalone server and also *on a cluster*. Also test the autofix service we have on a cluster.

**Comment by nqb** [comment]
I run a pipeline here to see if build passed: https://gitlab.com/inverse-inc/packetfence/-/pipelines/263867228.

If build passed, you should have a package ready for your tests on Debian and CentOS.

**Comment by lzammit** [comment]
I've tested out the upgrade so far and I get Mariadb errors:
````
Mar 12 13:46:07 unicorn mysqld: 2021-03-12 13:46:07 140469393680128 [Warning] Aborted connection 12967 to db: 'pf' user: 'pf' host: 'localhost' (Got an error reading communication packets)
Mar 12 13:46:07 unicorn mysqld: 2021-03-12 13:46:07 140469389686528 [Warning] Aborted connection 12968 to db: 'pf' user: 'pf' host: 'localhost' (Got an error reading communication packets)
Mar 12 13:46:19 unicorn mysqld: 2021-03-12 13:46:19 140469394908928 [Warning] Aborted connection 12974 to db: 'pf' user: 'pf' host: 'localhost' (Got an error reading communication packets)
Mar 12 13:46:19 unicorn mysqld: 2021-03-12 13:46:19 140469390608128 [Warning] Aborted connection 12975 to db: 'pf' user: 'pf' host: 'localhost' (Got an error reading communication packets)
````

````
MariaDB [(none)]> SELECT VERSION();
+-----------------+
| VERSION()       |
+-----------------+
| 10.2.37-MariaDB |
+-----------------+
````

````
MariaDB [pf]> select * from pf_version;
+--------+---------+
| id     | version |
+--------+---------+
| 655872 | 10.2.0  |
| 655881 | 10.2.9  |
+--------+---------+
````

````
[root@unicorn pf]# rpm -qa | grep -i mariadb
MariaDB-compat-10.2.37-1.el7.centos.x86_64
MariaDB-shared-10.2.37-1.el7.centos.x86_64
MariaDB-common-10.2.37-1.el7.centos.x86_64
MariaDB-client-10.2.37-1.el7.centos.x86_64
MariaDB-server-10.2.37-1.el7.centos.x86_64
````

**Comment by lzammit** [comment]
haproxy-admin errors also:
````
Mar 12 13:41:35 unicorn.inverse.ca haproxy[17526]: [WARNING] 070/134135 (17526) : parsing [/usr/local/pf/var/conf/haproxy-admin.conf:110] : a 'http-request' rule pla
Mar 12 13:41:35 unicorn.inverse.ca haproxy[17526]: [WARNING] 070/134135 (17526) : parsing [/usr/local/pf/var/conf/haproxy-admin.conf:111] : a 'http-request' rule pla
Mar 12 13:41:35 unicorn.inverse.ca haproxy[17526]: [ALERT] 070/134135 (17526) : Error(s) found in configuration file : /usr/local/pf/var/conf/haproxy-admin.conf
Mar 12 13:41:35 unicorn.inverse.ca haproxy[17526]: [ALERT] 070/134135 (17526) : Fatal errors found in configuration.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: Failed to start PacketFence HAProxy Load Balancer for the Admin GUI.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: Unit packetfence-haproxy-admin.service entered failed state.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: packetfence-haproxy-admin.service failed.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: packetfence-haproxy-admin.service holdoff time over, scheduling restart.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: Stopped PacketFence HAProxy Load Balancer for the Admin GUI.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: start request repeated too quickly for packetfence-haproxy-admin.service
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: Failed to start PacketFence HAProxy Load Balancer for the Admin GUI.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: Unit packetfence-haproxy-admin.service entered failed state.
Mar 12 13:41:35 unicorn.inverse.ca systemd[1]: packetfence-haproxy-admin.service failed.
````

File: lib/pf/dal.pm [hunk 1]
```diff
@@ -114,6 +114,7 @@ sub db_execute {
         unless ($sth && $sth->execute(@bind)) {
             my $err = $dbh->err;
             my $errstr = $dbh->errstr;
+	    print STDERR "$errstr\n";
```
Comment (fdurand) [inline lib/pf/dal.pm]: remove tab

**Comment by extrafu** [comment]
What's preventing us from merging that branch now?

**Comment by julsemaan** [comment]
Needs at least a rebase

**Comment by fdurand** [comment]
I still testing it. I have some weird issues with it.

**Comment by julsemaan** [comment]
Here is a list of the things we'd need in my opinion:
- [x] Update clustering install guide if necessary
- [x] Create update procedure for standalone server
- [x] Create update procedure for a cluster
- [x] Test galera-autofix with MariaDB 10.2
- [ ] Perform a lot of tests of the clustering layer to ensure there are no regressions (hard reboot, soft reboot, pkill -9, etc)

File: docs/cluster/cluster_setup.asciidoc [hunk 2]
```diff
@@ -66,34 +66,22 @@ NOTE: In this example, the database stack uses the native PacketFence https://ma
 
 CAUTION: Galera cluster is only supported in 3 nodes cluster and more (with an odd number of servers).
 
-First, you will need to install, *on each servers*, Percona Xtrabackup for the synchronization to work correctly.
+First, you will need to install, *on each servers*, MariaDB-Backup for the synchronization to work correctly.
```
Comment (extrafu) [inline docs/cluster/cluster_setup.asciidoc]: MariaDB-Backup -> Mariabackup

File: docs/installation/appendix.asciidoc [hunk 3]
```diff
@@ -38,9 +38,9 @@ The node view option shows all information contained in the node database table
   52:54:00:12:35:02|1|2008-10-23 17:32:16||||unreg||||2008-10-23 21:12:21|||||
 
 [appendix]
-=== Restoring a Percona XtraBackup dump
+=== Restoring a Percona XtraBackup or or MariaDB Backup dump
```
Comment (extrafu) [inline docs/installation/appendix.asciidoc]: "or or" -> "or"  and Mariabackup

File: docs/installation/appendix.asciidoc [hunk 4]
```diff
@@ -38,9 +38,9 @@ The node view option shows all information contained in the node database table
...
 
-If you enabled Percona XtraBackup for your nightly backup, you can use the following instructions to restore it. In this example we will be restoring `/root/backup/packetfence-db-dump-innobackup-2016-12-20_00h31.xbstream.gz`
+If you enabled Percona XtraBackup or MariaDB Backup for your nightly backup, you can use the following instructions to restore it. In this example we will be restoring `/root/backup/packetfence-db-dump-innobackup-2016-12-20_00h31.xbstream.gz`
```
Comment (extrafu) [inline docs/installation/appendix.asciidoc]: Mariabackup

File: docs/installation/appendix.asciidoc [hunk 5]
```diff
@@ -74,21 +74,38 @@ Debian:
...
+
   # service packetfence-mariadb stop
   # mv /var/lib/mysql /var/lib/mysql.bak
   # mkdir /var/lib/mysql
   # mv * /var/lib/mysql
   # chown -R mysql: /var/lib/mysql
   # service packetfence-mariadb start
 
+
+For MariaDB-backup:
```
Comment (extrafu) [inline docs/installation/appendix.asciidoc]: Mariabackup

File: addons/backup-and-maintenance.sh [hunk 6]
```diff
@@ -110,15 +110,15 @@ should_backup(){
 }
 
 backup_db(){
-    # Check to see if Percona XtraBackup is installed
-    if hash innobackupex 2>/dev/null; then
-        echo -e "Percona XtraBackup is available. Will proceed using it for DB backup to avoid locking tables and easier recovery process. \n"
-        PERCONA_XTRABACKUP_INSTALLED=1
+    # Check to see if MariaDB-Backup is installed
```
Comment (extrafu) [inline addons/backup-and-maintenance.sh]: Mariabackup

File: addons/backup-and-maintenance.sh [hunk 7]
```diff
@@ -110,15 +110,15 @@ should_backup(){
...
+    if hash mariabackup 2>/dev/null; then
+        echo -e "MariaDB-Backup is available. Will proceed using it for DB backup to avoid locking tables and easier recovery process. \n"
```
Comment (extrafu) [inline addons/backup-and-maintenance.sh]: Mariabackup

File: addons/backup-and-maintenance.sh [hunk 8]
```diff
@@ -132,23 +132,23 @@ backup_db(){
...
             echo "----- Backup started on `date +%F_%Hh%M` -----" >> /usr/local/pf/logs/innobackup.log
             INNO_TMP="/tmp/pf-innobackups"
             mkdir -p $INNO_TMP
             if [ $MARIADB_LOCAL_CLUSTER -eq 1 ]; then
-                innobackupex --defaults-file=/usr/local/pf/var/conf/mariadb.conf --user=$REP_USER --password=$REP_PWD  --no-timestamp --stream=xbstream --tmpdir=$INNO_TMP $INNO_TMP 2>> /usr/local/pf/logs/innobackup.log | gzip - > $BACKUP_DIRECTORY/$BACKUP_DB_FILENAME-innobackup-`date +%F_%Hh%M`.xbstream.gz
+                mariabackup --defaults-file=/usr/local/pf/var/conf/mariadb.conf --user=$REP_USER --password=$REP_PWD  --no-timestamp --stream=xbstream --tmpdir=$INNO_TMP --backup 2>> /usr/local/pf/logs/innobackup.log | gzip - > $BACKUP_DIRECTORY/$BACKUP_DB_FILENAME-innobackup-`date +%F_%Hh%M`.xbstream.gz
             elif [ $MARIADB_REMOTE_CLUSTER -eq 1 ]; then
-                echo "innobackupex can't backup remote databases, uninstall Percona XtraBackup and retry"
+                echo "mariabackup can't backup remote databases, uninstall MariaDB-backup and retry"
```
Comment (extrafu) [inline addons/backup-and-maintenance.sh]: Mariabackup

File: docs/installation/appendix.asciidoc [hunk 9]
```diff
@@ -74,21 +74,38 @@ Debian:
   # apt-get install qpress
   # for i in $(find -name "*.qp"); do qpress -vd $i $(dirname ${i}) && rm -f $i; done
 
-Next, apply the innodb log:
+Next, apply the innodb log (for XtraBackup):
 
   # innobackupex --apply-log ./
 
+Next, apply the innodb log (for MariaDB-backup):
```
Comment (extrafu) [inline docs/installation/appendix.asciidoc]: Mariabackup

**Pull request merged**